### PR TITLE
Improvement of the DRAMPower library

### PR DESCRIPTION
**Pull request by fsprado**
## CommandAnalysis:

A self-refresh command is modelled as an auto-refresh command followed by the trigger of the DRAM internal self-timed refresh, disabling the clock pin, if the self-refresh period is greater than a refresh period. The DRAMPower uses mainly three variables to calculate the energy consumed during a self-refresh period: [*sref\_cycle*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-786e44d4eec07c365c733dc9c99a83edR174), [*sref\_ref\_act\_cycles*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-786e44d4eec07c365c733dc9c99a83edR109) and [*sref\_ref\_pre\_cycles*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-786e44d4eec07c365c733dc9c99a83edR113)). The variable *sref\_cycle* represents the cycles after triggering the DRAM internal self-timed refresh, and the others
represent the active and precharged auto-refresh cycles. A self-refresh period can be shorter, equal or longer than a refresh period, hence, it is necessary while calculating the energy consumed during a self-refresh, which variables must be used. By using the windowing system, there is the possibility of a window calculation ending during a self-refresh phase. Hence, these three variables must have their "window" versions ([*sref\_cycle\_window*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-786e44d4eec07c365c733dc9c99a83edR163), [*sref\_ref\_act\_cycles\_window*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-786e44d4eec07c365c733dc9c99a83edR111) and [*sref\_ref\_pre\_cycles\_window*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-786e44d4eec07c365c733dc9c99a83edR115)) to calculate correctly the energy consumed during a self-refresh, even if this period is divided into windows. Besides, prior to the implementation of the windowing system, the self-refresh energy consumed was calculated based on its start and end cycles. Introducing the windowing for simulator coupling (e.g. with gem5 or DRAMSys), the evaluation of commands timings and memory states transitions remained the same and it would be correct if the window size was always longer than any self-refresh period. However, there is one [ if-clause](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-a4d7a2a5d67f32407ca63d873a3d5b83R463) when a self-refresh period ends that depens on its start cycle (the variable *sref\_cycle*). When the window size is shorter than a self-refresh period, due to the window energy calculation in the middle of a self-refresh period, the [*clearStats*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-a4d7a2a5d67f32407ca63d873a3d5b83R74) function is called and hence this variable is assigned to the current timestamp. This assignment can introduce a conditional jump error, in other words, a energy calculation error. By adding the variable *sref\_cycle\_window*, which is assigned [here](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-a4d7a2a5d67f32407ca63d873a3d5b83L110) to the current timestamp instead of the variable *sref\_cycle*, this problem is solved. Now *clearStats* changes just the new variable, keeping the *sref\_cycle* value correct during each self-refresh period.

The variable [*last\_pre\_cycle*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-786e44d4eec07c365c733dc9c99a83edR189) can be greater than the current timestamp if a window calculation stops in the middle of a refresh period. Hence, the function [*clearStats*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-a4d7a2a5d67f32407ca63d873a3d5b83R128) must assign this variable to the current timestamp only if it is not greater than the current timestamp, otherwise the energy consumed by precharged cycles will be calculated wrongly.

When a WRA/RDA command is issued, the function [*getCommands*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-a4d7a2a5d67f32407ca63d873a3d5b83R153) will divide these command in two: a WR/RD command and a PRE command. By using the windowing system, a window energy would be calculated including these commands even if the PRE command issue timestamp was greater than the current timestamp, what would result in a energy calculation error. By adding the vector [*next\_command\_cmd\_list*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-786e44d4eec07c365c733dc9c99a83edR149) , every time a command whose issue timestamp is greater than the current timestamp is issued, this scenario is avoided, and this command will be correctly included in the energy calculation of the next window.

The number of active cycles during an auto-refresh is defined according to *tRFC* (Row Refresh Cycle Time) and to *tRP* (Row Precharge Time), parameters of the memory specification . Therefore, the values of the variables [*spup\_act\_cycles*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-786e44d4eec07c365c733dc9c99a83edR118) and *sref\_ref\_act\_cycles* must be *tRFC* - *tRP* at most. Due to this condition,  the maximum possible value of the variable [*spup\_act*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-a4d7a2a5d67f32407ca63d873a3d5b83L489)  was changed to *tRFC* - *tRP* instead of just *tRFC*.

## MemoryPowerModel:

After implementing the windowing system, the variables [*total\_energy*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-14dcb72ab3729df9e82afd0abe0f5056R97) and [*average\_power*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-14dcb72ab3729df9e82afd0abe0f5056R148), which belong to the Energy and Power structures, respectively, were used to represent either a time window energy or the total energy. Also the variable [*total\_cycles*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-14dcb72ab3729df9e82afd0abe0f5056L65) was only used to calculate the cycles of a window because it was reset everytime a window energy was calculated. Just to make the code easier to understand, three variable were added: [*window\_energy*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-14dcb72ab3729df9e82afd0abe0f5056R100), [*window\_average\_power*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-14dcb72ab3729df9e82afd0abe0f5056R151) and [*window\_cycles*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-14dcb72ab3729df9e82afd0abe0f5056R67). Now the [old](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-32cd4e872bf7e18e4ab20c4e1842fcd4R255) variables are related to the energy calculation of the entire simulation and the [new](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-32cd4e872bf7e18e4ab20c4e1842fcd4R141) ones are related to the energy calculation of a time window.

## LibDRAMPower:

Previously it was necessary to call more than one function to perform the complete energy calculation in the used simulator. The functions[ *updateCounters*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-a5608d1595ab7016e8281f884560d2b9R59)  and *calcEnergy* were called to calculate the total energy and besides these, to calculate a time window energy, the functions [*clearCounters* ](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-a5608d1595ab7016e8281f884560d2b9R85)and [*doCommand*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-a5608d1595ab7016e8281f884560d2b9R53) were required as well. Now it is needed just one function call to calculate the energy consumed: [*calcEnergy*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-a5608d1595ab7016e8281f884560d2b9R65) to calculate the total energy, and [*calcWindowEnergy*](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-a5608d1595ab7016e8281f884560d2b9R71) to calculate a time window energy.

In fact the library interface is reduced to only 5 integral function calls:

``` c++
// Send DRAMPower Library a specific command and timestamp
doCommand(...);
// Compute energy until a specific timestamp in order to support windowing 
calcWindowEnergy(...);
// Finishes the simulation and compute the TOTAL results including the last window.
calcEnergy();
// Query energy and power values for windows and total.
getEnergy();
getPower();
```

The file [window_example.cc](https://github.com/ravenrd/DRAMPower/compare/master...fsprado:master#diff-bc7f00eb9b8a92f0c4724c821f1239ffR1)  shows an example for an easy handling of the library. Also, the test [*test\_window\_example*](https://github.com/ravenrd/DRAMPower/pull/34/files#diff-3eb4d995269f8dd7de58cb236d151d7bR167) was added to the file *test.py*, in which the window_example output should match the DRAMPower output with the command trace window.trace.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -115,21 +115,21 @@ make -j4
 
 This should show the following compilation message on the screen:
 ```
-g++ -O -W -pedantic-errors -Wextra -Werror -Wformat -Wformat-nonliteral -Wpointer-arith -Wcast-align -Wconversion  -g  -std=c++98 -MMD -MF src/xmlparser/MemSpecParser.d -iquote src -o src/xmlparser/MemSpecParser.o -c src/xmlparser/MemSpecParser.cc
```
Comment (Sv3n) [inline README.md]: In the next iteration of the readme we should really get rid of this list...it's fairly pointless :)

File: src/MemoryPowerModel.cc [hunk 2]
```diff
@@ -65,7 +65,7 @@ void MemoryPowerModel::power_calc(const MemorySpecification& memSpec,
   energy.pre_stdby_energy    = 0.0;
   energy.idle_energy_act     = 0.0;
   energy.idle_energy_pre     = 0.0;
-  energy.total_energy        = 0.0;
+  energy.window_energy        = 0.0;
```
Comment (Sv3n) [inline src/MemoryPowerModel.cc]: -1 space :)

File: src/MemoryPowerModel.h [hunk 3]
```diff
@@ -62,7 +62,9 @@ class MemoryPowerModel {
                           double spup_ref_pre_cycles,
                           double clk);
 
-  int64_t total_cycles;
+  int64_t total_cycles = 0;
```
Comment (Sv3n) [inline src/MemoryPowerModel.h]: Line 65 and 68 break the builds for somewhat older versions of gcc: https://travis-ci.org/ravenrd/DRAMPower/jobs/149783410

Can you please move the initialization of these variables to a different place to fix that?

File: test/libdrampowertest/commands.trace [hunk 4]
```diff
@@ -16,4 +16,4 @@
 210,RDA,0
 232,ACT,4
 247,WRA,4
-248,PDN_F_ACT,3
+265,PDN_F_PRE,3
```
Comment (Sv3n) [inline test/libdrampowertest/commands.trace]: Why has this command changed?

File: test/libdrampowertest/window_example.cc [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+#include <string>
+#include "libdrampower/LibDRAMPower.h"
+#include "xmlparser/MemSpecParser.h"
+
+using namespace std;
+using namespace Data;
+
+int main(int argc, char* argv[])
+{
+        assert(argc == 2);
```
Comment (Sv3n) [inline test/libdrampowertest/window_example.cc]: The indentation in this file is a bit wobbly.

File: test/libdrampowertest/window_example.cc [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        libDRAMPower test = libDRAMPower(memSpec, 0);
+	int i = 0;
+	unsigned long windowSize = 70;
+
+        test.doCommand(MemCommand::ACT,0,25);
+        test.doCommand(MemCommand::RD,0,35);
+	test.doCommand(MemCommand::PRE,0,45);
+        test.doCommand(MemCommand::ACT,4,55);
+        test.doCommand(MemCommand::WR,4,65);
+	test.calcWindowEnergy(++i*windowSize);
```
Comment (Sv3n) [inline test/libdrampowertest/window_example.cc]: clang complaints here:
```c++
 test.calcWindowEnergy(++i*windowSize);
```

```
./test/libdrampowertest/window_example.cc:77:27: error: implicit conversion changes signedness: 'unsigned long' to 'int64_t' (aka 'long long') [-Werror,-Wsign-conversion]
        test.calcWindowEnergy(++i*windowSize);
        ~~~~                  ~~~^~~~~~~~~~~
./test/libdrampowertest/window_example.cc:77:24: error: implicit conversion changes signedness: 'int' to 'unsigned long' [-Werror,-Wsign-conversion]
        test.calcWindowEnergy(++i*windowSize);
```

**Comment by Sv3n** [comment]
I've had a look at the pull request. Overall it looks fine, I think it's a good addition to the code. The biggest issue I have with it currently is that it breaks the CI tests, but that can be fixed relatively easily. You can have a look at my comments for details.

File: test/libdrampowertest/commands.trace (hunk 4 above)
Comment (fsprado) [inline test/libdrampowertest/commands.trace]: The last command in the file commands.trace was different from the last command that is done in the file lib_test.cc.

**Comment by fsprado** [comment]
Please wait before merging the pull request. I have made some changes and I will commit them soon.

**Comment by Sv3n** [comment]
Ok! I was waiting for the build to be fixed anyway :)

**Comment by Sv3n** [comment]
Thanks for fixing the build. I'll do a final check this evening (build of merge on local machine). If it works as expected, then I'll push the merge button and close this pull request.

**Pull request merged**